flows/atd-knack-services/corridor-retiming.py

```diff
@@ -0,0 +1,145 @@
+--name "Knack Services: Corridor Retiming" --pool atd-data-03 \
+--cron "45 11 * * *" -q default -sb github/atd-prefect-main-branch \
+-o "deployments/knack-services-corridor-retiming.yaml" --skip-upload
+ 
+$ prefect deployment apply deployments/knack-services-corridor-retiming.yaml
+"""
+
+import os
+
+import docker
+import prefect
+from datetime import datetime, timedelta
+
+# Prefect
+from prefect import flow, task, get_run_logger
+from prefect.blocks.system import JSON
+
+
+# Docker settings
+docker_env = "production"
+docker_image = f"atddocker/atd-knack-services:{docker_env}"
+
+
+@task(
+    name="get_env_vars",
+    retries=10,
+    retry_delay_seconds=timedelta(seconds=15).seconds,
+)
+def get_env_vars(json_block):
+    # Environment Variables stored in JSON block in Prefect
+    return JSON.load(json_block).dict()["value"]
+
+
+@task(
+    name="pull_docker_image",
+    retries=1,
+    retry_delay_seconds=timedelta(minutes=5).seconds,
+)
+def pull_docker_image():
+    client = docker.from_env()
+    client.images.pull(docker_image)
+    return True
+
+
+# Determine what date to append to knack scripts arguments
+@task(
+    name="determine_date_args",
+    retries=1,
+    retry_delay_seconds=timedelta(minutes=5).seconds,
+)
+def determine_date_args(environment_variables, commands):
+    # Completely replace data on 15th day of every month,
+    # to catch records potentially missed by incremental refreshes
+    output = []
+    if datetime.today().day == 15:
+        for c in commands:
+            output.append(f"{c} -d 1970-01-01")
+        return output
+
+    prev_exec = environment_variables["PREV_EXEC"]
+    for c in commands:
+        output.append(f"{c} -d {prev_exec}")
+    return output
+
+
+@task(
+    name="docker_commands",
+    retries=1,
+    retry_delay_seconds=timedelta(minutes=5).seconds,
+)
+def docker_commands(environment_variables, commands, logger):
+    # list ex: ["atd-knack-services/services/records_to_postgrest.py -a {app_name} -c {container} -d {date_filter}"]
+    for c in commands:
+        response = (
+            docker.from_env()
+            .containers.run(
+                image=docker_image,
+                working_dir=None,
+                command=f"python {c}",
+                environment=environment_variables,
+                volumes=None,
+                remove=True,
+                detach=False,
+                stdout=True,
+            )
+            .decode("utf-8")
+        )
+        logger.info(response)
+    return response
+
+
+@task(
+    name="update_exec_date",
+    retries=10,
+    retry_delay_seconds=timedelta(seconds=15).seconds,
+)
+def update_exec_date(env_block_name):
+    # Update our JSON block with the date of last successful flow execution
+    block = JSON.load(env_block_name)
+    block.value["PREV_EXEC"] = datetime.today().strftime("%Y-%m-%d")
+    block.save(name=env_block_name, overwrite=True)
```

Comment:
This is game changing. I didn't realize we had such a readily available mechanism to save state in the prefect ecosystem. I can see a world where we track moped test deployments in a block and use it to clean up after us automatically after some fixed duration.

Reply:
credit for this all goes to @Charlie-Henry